sw/device/tests/flash_ctrl_idle_low_power_test.c

```diff
@@ -0,0 +1,117 @@
+  dif_pwrmgr_t pwrmgr;
+  dif_rstmgr_t rstmgr;
+  dif_aon_timer_t aon;
+
+  CHECK_DIF_OK(dif_flash_ctrl_init_state(
+      &flash, mmio_region_from_addr(TOP_EARLGREY_FLASH_CTRL_CORE_BASE_ADDR)));
+  CHECK_DIF_OK(dif_pwrmgr_init(
+      mmio_region_from_addr(TOP_EARLGREY_PWRMGR_AON_BASE_ADDR), &pwrmgr));
+  CHECK_DIF_OK(dif_rstmgr_init(
+      mmio_region_from_addr(TOP_EARLGREY_RSTMGR_AON_BASE_ADDR), &rstmgr));
+  CHECK_DIF_OK(dif_aon_timer_init(
+      mmio_region_from_addr(TOP_EARLGREY_AON_TIMER_AON_BASE_ADDR), &aon));
+
+  CHECK_DIF_OK(dif_aon_timer_watchdog_stop(&aon));
+  CHECK_DIF_OK(dif_pwrmgr_set_request_sources(&pwrmgr, kDifPwrmgrReqTypeReset,
+                                              kDifPwrmgrResetRequestSourceTwo,
+                                              kDifToggleEnabled));
+
+  dif_rstmgr_reset_info_bitfield_t rstmgr_reset_info;
+  CHECK_DIF_OK(dif_rstmgr_reset_info_get(&rstmgr, &rstmgr_reset_info));
+  CHECK_DIF_OK(dif_rstmgr_reset_info_clear(&rstmgr));
+
+  uint32_t address = flash_ctrl_testutils_data_region_setup(
+      &flash, kRegionBasePageIndex, kFlashDataRegion, kRegionSize);
+
+  uint32_t readback_data[kNumWords];
+
+  if (rstmgr_reset_info == kDifRstmgrResetInfoPor) {
+    // Create data. Simple count for data is sufficient which will
+    // be different than the 0xFFFFFFFF that is created
+    // with an erase.
+    uint32_t data[kNumWords];
+    for (int i = 0; i < kNumWords; ++i) {
+      data[i] = i;
+    }
+
+    // Erasing the page and writing data to it followed
+    // by a read back and compare to sanity check basic operation.
+    CHECK(flash_ctrl_testutils_erase_and_write_page(
+              &flash, address, kPartitionId, data,
+              kDifFlashCtrlPartitionTypeData, kNumWords) == 0);
+    CHECK(flash_ctrl_testutils_read_page(
+              &flash, address, kPartitionId, readback_data,
+              kDifFlashCtrlPartitionTypeData, kNumWords, 0) == 0);
+    CHECK_BUFFER(data, readback_data, kNumWords);
+
+    // Setting up low power hint and starting watchdog timer followed by
+    // a flash operation (page erase) and WFI. This will create a reset
+    // at some time following the flash operation.
+    // On the second boot we can confirm that the there was no wake
+    // from low power state and also check that the erase
+    // was successful.
+    pwrmgr_testutils_enable_low_power(&pwrmgr, kDifPwrmgrWakeupRequestSourceTwo,
+                                      0);
+
+    CHECK_DIF_OK(dif_aon_timer_watchdog_stop(&aon));
+    CHECK_DIF_OK(
+        dif_aon_timer_watchdog_start(&aon, 0xffffffff, 0x40, false, false));
+
+    dif_flash_ctrl_transaction_t transaction = {
+        .byte_address = address,
+        .op = kDifFlashCtrlOpPageErase,
+        .partition_type = kDifFlashCtrlPartitionTypeData,
+        .partition_id = kPartitionId,
+        .word_count = 0x0};
+
+    CHECK_DIF_OK(dif_flash_ctrl_start(&flash, transaction));
+    wait_for_interrupt();
+
```

Comment:
If the flash operation got started then the entry to low power will not happen, which you can confirm via WAKE_INFO.ABORT as mentioned above. This also means the CPU will remain active and you could wait for flash to complete (if there is a flash_ctrl bit that flips once it completes, so spin_wait on a csr read), and check the expected result right once it does... no need to wait for a watchdog...

Reply:
Thanks, can you explain more on this?
Once the low power hint is set, I start the flash operation and then
call wait_for_interrupt. If it is working as expected this should NOT enter
low power, however I believe the CPU still halts on a WFI instruction. Any
code after the wait_for_interrupt isn't executed (I gave it a try!). At this point
the CPU needs to be brought out of this halted condition, hence the watchdog.

Also, the WAKE_INFO.ABORT register is set but only after a wake event? I have added it to 
the reboot branch and check for the presence of it as an extra check.

Reply:
The AON timer watchdog should cause a bark interrupt first before the bite, which causes a reset. You have not programmed the PLIC (or defined the ISR routine to service the bark interrupt. Hence it proceeds to a bit, causing a reset. That is not actually needed. Just configure the bark irq to interrupt the CPU - that is sufficient to get it out of WFI. You can read the reset info again to verify we are still in POR. As such, this test does not need to cycle through a reset and reboot.

Reply:
Thanks @sriyerg, sounds good, I have changed to use the bark interrupt.